### ISSUE TITLE
Support predefined configurations/configurations

### ISSUE THREAD

**Issue by Zingam** [microsoft/vscode-cmake-tools#1011]
https://github.com/microsoft/vscode-cmake-tools/issues/1003#issuecomment-574020010

Instead of just letting the user to select: build variant, kit, etc. the extension can have named configurations, including predefined configurations like in VS2019. These can be extended to support configurations defined in a CMakeSettings.json file like in VS2019.

**Comment by farrrb** [comment]
I'd really like this, because I need configurations that bundle variants and kits together for cross-compilation.

e.g. 
Build:
1. **STM32F205** Release µC always with arm-none-eabi toolchain 
but build: 
2. **Unit Tests** Debug always with clang on windows.

**Comment by bobbrow** [comment]
This feels like a duplicate of #529.  Correct me if I'm wrong, but I think that support for CMakeSettings.json would address this.

**Comment by bobbrow** [comment]
Thinking about this more, I believe there might be more to it than that.  Though I'm unsure of what exactly you would like that is different from the default variants.  Based on the comment from the linked thread it appears you would like to have different default `buildDirectory`s based on the Kit/generator.

**Comment by farrrb** [comment]
I think the idea of a CMakeSettings.json (as it is in Visual Studio right now) is kind of orthogonal to the variants concept of cmake-tools (in VS Code). 

Variants are pretty flexible and they enable you to combine everything with everything (like a big multi-dimensional matrix). 
The idea behind the CMakeSettings.json is to provide fixed configurations (build type, variant, cmake toolchain file, ....) .

I think it will be pretty difficult to support both concepts (variants + kits OR configurations with toolchain files). 

Both approaches would work for my use case which is just to fixate some configuration parameters (like build type, build directory, generator, ...) to a fixed toolchain/kit (for cross-compiling or native compiling). I would prefer to use the same concept in VS Code and in "normal" VS, just to keep the complexity low.

**Comment by orion160** [comment]
Adding this feature to vscode would bring more operability to c++ projects using microsoft tools. 
The use case that should be common is to provide a crossplatform enviroment. The current limitation is that there is no common aproach to support developers in diferent OS (aka. there is no visual studio for linux -> so a person in the development team must use microsoft windows + Visual Studio).

So in short words, the way to develop with cmake and Visual Studio, is to EDIT in Visual Studio and launch a remote build.

So what will bring this feature is to edit in a crossplatform code editor as vscode and launch local builds using the CmakeSettings.json.

**Comment by Zingam** [comment]
@orion160 Just as side note AFAIK both Visual Studio and I guess this extension will be moving to support CMake presets and CMakeSettings.json will be probably deprecated.

I really wish tighter integration between Visual Studio and Visual Studio Code/CMake tools to support cross-platform development seamlessly. There are number of areas they defer like the VSC extension doesn't support install and debug yet as well passing command line arguments via launch.vs.json incompatible or at least it used to be.

**Comment by orion160** [comment]
This seems to be a duplicate of #529

**Comment by Zingam** [comment]
Actually not quite. I am not sure how relevant is this stuff currently.
